### PR TITLE
Switch to using deb packaging and systemd

### PR DESCRIPTION
**Pull request by rtyley**
Further to #1367, this removes the old `upstart` configuration in favour of `systemd` and means we
can use modern [Amigo-generated AMIs](https://amigo.gutools.co.uk/recipes/xenial-membership).

The configuration here is based upon the similar config found in Contributions frontend:

* https://github.com/guardian/contributions-frontend/blob/9830dbf1/build.sbt#L77-L101
* https://github.com/guardian/contributions-frontend/blob/9830dbf1/cloud-formation/cfn.yml#L105-L156

The relevant changes in Contributions that resulted in this config include:

* guardian/contributions-frontend@ec38edc
* guardian/contributions-frontend#198

cc @guardian/membership-and-subscriptions

### CODE REVIEW

File: cloud-formation/membership-app.cf.json [hunk 1]
```diff
@@ -102,35 +102,28 @@
...
+                        "CONF_DIR=/etc/frontend",
 
-                        "/opt/features/ssh-keys/initialise-keys-and-cron-job.sh -l -b github-team-keys -t Membership-and-Subscriptions || true",
-
-                        "adduser --system --home /membership --disabled-password membership",
-
-                        {"Fn::Join":["", ["aws --region ", { "Ref": "AWS::Region" }, " s3 cp s3://membership-dist/", { "Ref" : "Stack" }, "/", { "Ref" : "Stage" }, "/frontend/frontend-1.0-SNAPSHOT.tgz /tmp"]]},
-
-                        "tar xvz -C /membership -f /tmp/frontend-1.0-SNAPSHOT.tgz",
+                        {"Fn::Join":["", ["aws --region ", { "Ref": "AWS::Region" }, " s3 cp s3://membership-dist/", { "Ref" : "Stack" }, "/", { "Ref" : "Stage" }, "/frontend/frontend_1.0-SNAPSHOT_all.deb /tmp"]]},
```
Comment (paulbrown1982) [inline cloud-formation/membership-app.cf.json]: There's an Fn::Sub command now which lets you reference Refs via just $variable inside the string.

File: cloud-formation/membership-app.cf.json [hunk 2]
```diff
@@ -102,35 +102,28 @@
...
+                        "dpkg -i /tmp/frontend_1.0-SNAPSHOT_all.deb",
+                        "chown frontend /etc/gu/membership.private.conf",
+                        "chmod 0600 /etc/gu/membership.private.conf",
+
                         {"Fn::Join": ["", ["sed -i",
                             " -e \"s/__DATE/$(date +%F)/\"",
-                            " -e \"s/__BUILD/$(cat $CONF_DIR/build.txt)/\"",
                             " -e 's/__STAGE/", { "Ref": "Stage" }, "/'",
                             " $CONF_DIR/logger.conf"]]},
                         {"Fn::Join": [" ", ["python awslogs-agent-setup.py -nr", { "Ref": "AWS::Region" }, "-c $CONF_DIR/logger.conf" ]]},
```
Comment (paulbrown1982) [inline cloud-formation/membership-app.cf.json]: I vaguely recall @alexduf using a different kind of collector (non-python based) for contributions-frontend?

Comment (rtyley) [inline cloud-formation/membership-app.cf.json]: Hmm, looks they _are_ still using this Python agent on Contributions:

https://github.com/guardian/contributions-frontend/blob/9830dbf1/cloud-formation/cfn.yml#L151-L156

**Comment by paulbrown1982** [comment]
It's great to get rid of the old gu-membership AMI baking stuff - I should have deleted that when we started using the Tools AMIs. However, moving to Amigo is totally the way to go. It would be good to get help with moving the other apps sometime soon too. 👍

File: cloud-formation/membership-app.cf.json (hunk 1 above)
Comment (rtyley) [inline cloud-formation/membership-app.cf.json]: Wow, that makes everything look much better! Hopefully the changes I made in 1318ec5 are all valid.

**Comment by philwills** [comment]
You can actually do one big multiline `Sub`, like https://github.com/guardian/contributions-frontend/blob/simpler-cloudformation/cloud-formation/cfn.yml#L106

**Comment by rtyley** [comment]
> You can actually do one big multiline Sub

I think that's probably one of the several things that works better in YAML. For instance if you look at http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-sub.html#d0e160110 :

image

...so yeah, we should probably switch to using YAML - I'll leave that to another PR tho!

**Comment by philwills** [comment]
Oops, sorry, didn't realise you weren't already on YAML.

**Comment by philwills** [comment]
Also, these are the type of stats that make me happy:

image

**Comment by rtyley** [comment]
All went well, thanks to @svillafe for helping!

image

[Amiable is also happy](https://amiable.gutools.co.uk/instanceAMIs?stack=membership):

image


I can successfully SSH to the boxes, so Amigo's [s3-ssh-keys](https://amigo.gutools.co.uk/roles#s3-ssh-keys) is working, and getting logs on the box with journalctl also looks good:

```
$ journalctl -u frontend
-- Logs begin at Thu 2016-11-17 10:36:57 UTC, end at Thu 2016-11-17 10:52:01 UTC. --
Nov 17 10:38:19 ip-10-248-132-102 systemd[1]: Started Membership Frontend service.
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:28,525 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Could NOT find resource [logback.groovy]
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:28,525 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Found resource [logback-test.xml] at [file:/usr/share/frontend/conf/logback-test.xml]
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,098 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - debug attribute not set
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,253 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [ch.qos.logback.core.ConsoleAppender]
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,255 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [CONSOLE]
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,629 |-WARN in ch.qos.logback.core.ConsoleAppender[CONSOLE] - This appender no longer admits a layout as a sub-component, set an encoder instead.
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,629 |-WARN in ch.qos.logback.core.ConsoleAppender[CONSOLE] - To ensure compatibility, wrapping your layout in LayoutWrappingEncoder.
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,629 |-WARN in ch.qos.logback.core.ConsoleAppender[CONSOLE] - See also http://logback.qos.ch/codes.html#layoutInsteadOfEncoder for details
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,629 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [ch.qos.logback.core.FileAppender]
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,630 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [FILE]
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,631 |-WARN in ch.qos.logback.core.FileAppender[FILE] - This appender no longer admits a layout as a sub-component, set an encoder instead.
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,631 |-WARN in ch.qos.logback.core.FileAppender[FILE] - To ensure compatibility, wrapping your layout in LayoutWrappingEncoder.
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,631 |-WARN in ch.qos.logback.core.FileAppender[FILE] - See also http://logback.qos.ch/codes.html#layoutInsteadOfEncoder for details
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,631 |-INFO in ch.qos.logback.core.FileAppender[FILE] - File property is set to [logs/acceptance-test.log]
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,632 |-INFO in ch.qos.logback.classic.joran.action.RootLoggerAction - Setting level of ROOT logger to INFO
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,632 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [CONSOLE] to Logger[ROOT]
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,633 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [FILE] to Logger[ROOT]
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,633 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - End of configuration.
Nov 17 10:38:29 ip-10-248-132-102 frontend[1606]: 10:38:29,634 |-INFO in ch.qos.logback.classic.joran.JoranConfigurator@156b88f5 - Registering current configuration as safe fallback point
```

File: frontend/conf/upstart.conf [hunk 3]
```diff
@@ -1,27 +0,0 @@
...
-stop on runlevel [016]
-
-# NB: setuid is only supported on recent versions of upstart - i.e.
-#  on ubuntu not on amazon linux
-
-setuid membership
-
-chdir /membership
-
-limit nofile 65536 65536
```
Comment (rtyley) [inline frontend/conf/upstart.conf]: You ultimately fixed this with https://github.com/guardian/amigo/pull/94, is that right?

Comment (johnduffell) [inline frontend/conf/upstart.conf]: yes!  all good now 👍

**Pull request merged**